lua/tint.lua

```diff
@@ -169,12 +176,79 @@ local function setup_namespaces()
     __.tint_ns = vim.api.nvim_create_namespace("_tint_dim")
   end
 
-  for hl_group_name, hl_def in pairs(get_global_highlights()) do
+  for hl_group_name, hl_def in pairs(get_highlights(0)) do
     -- Ensure we only have valid keys copied over
     hl_def = ensure_valid_hl_keys(hl_def)
     set_default_ns(hl_group_name, hl_def)
-    set_tint_ns(hl_group_name, hl_def)
+    set_tint_ns(hl_group_name, hl_def, __.tint_ns)
+  end
+end
+
+---@param hl_ns_id number
+---@param suffix string
+local function add_namespace(hl_ns_id, suffix)
+  local ns_name = "tint_ns_" .. tostring(suffix)
+  local ns_id = vim.api.nvim_create_namespace("_tint_dim_" .. tostring(suffix))
+  __[ns_name] = ns_id
+
+  -- first copy over global tint namespace
+  for hl_group_name, hl_def in pairs(get_highlights(__.tint_ns)) do
+    vim.api.nvim_set_hl(ns_id, hl_group_name, hl_def)
+  end
+  -- create tinted highlights for extra namespace
+  for hl_group_name, hl_def in pairs(get_highlights(hl_ns_id)) do
+    -- Ensure we only have valid keys copied over
+    hl_def = ensure_valid_hl_keys(hl_def)
+    set_tint_ns(hl_group_name, hl_def, ns_id)
+  end
+end
+
+---@param winid number
+---@return number
+local function get_untint_ns_id(winid)
+  local untint_ns_id = vim.w[winid].untint_ns_id
+  if untint_ns_id == nil then
+    untint_ns_id = vim.api.nvim_get_hl_ns and vim.api.nvim_get_hl_ns({ winid = winid })
+    if untint_ns_id == nil or untint_ns_id < 0 then
+      untint_ns_id = 0
+    end
+    vim.api.nvim_win_set_var(winid, "untint_ns_id", untint_ns_id)
   end
+  return untint_ns_id
+end
+
+---@param winid number
+---@return number
+local function get_tint_ns_id(winid)
+  local untint_ns_id = get_untint_ns_id(winid)
+
+  local tint_ns_id
+  if untint_ns_id == 0 then
```

Comment:
_nit_: Can simplify things here a bit by doing:

```lua
if untint_ns_id == 0 or not __.user_config.should_create_extra_tint(winid, untint_ns_id, ns_suffix) then
```

to make the function logic a bit more readable here - either we are just using the global namespace or we are using the user-enabled stuff.

Reply:
I think you are right and it is better to make it more readable. For your suggestion I would need the ns_suffix always, so i decided to go a different way...to accomplish readability.
The result uses the "should_create_extra_tint" method only if no __.tint_ns_<original> id exists already. To simplify a little more, i renamed the defaults tint_ns to __.tint_ns_0